x-pack/plugins/encrypted_saved_objects/server/saved_objects/index.ts

```diff
@@ -96,6 +106,38 @@ export function setupSavedObjects({
           )) as T,
         };
       },
+
+      createPointInTimeFinderAsInternalUser: async <T = unknown, A = unknown>(
+        findOptions: SavedObjectsCreatePointInTimeFinderOptions,
+        dependencies?: SavedObjectsCreatePointInTimeFinderDependencies
+      ): Promise<ISavedObjectsPointInTimeFinder<T, A>> => {
+        const [internalRepository, typeRegistry] = await internalRepositoryAndTypeRegistryPromise;
+        const finder = internalRepository.createPointInTimeFinder<T, A>(findOptions, dependencies);
+        const finderAsyncGenerator = finder.find();
+
+        async function* encryptedFinder() {
+          for await (const res of finderAsyncGenerator) {
+            const encryptedSavedObjects = await pMap(res.saved_objects, async (savedObject) => ({
```

Comment:
__potential issues:__ 

1. How does `pMap` work when we have thousands of saved objects? Does it try to process everything in parallel? If so, I'm afraid it can quickly make Kibana unusable due to high cost of concurrent decryption operations. Did you have a chance to stress-test this a bit? If there a way to control how much tasks it can handle in parallel, the tests can give us a ballpark number.

2. What is the consumer (alerting) expectation if some of the saved objects cannot be decrypted (e.g. wrong encryption key)? How does API behave in this case? I think it makes sense to have a bunch of integration tests for this new API similar (but not limited to) to what we have for [ getDecryptedAsInternalUser](https://github.com/elastic/kibana/blob/main/x-pack/test/encrypted_saved_objects_api_integration/tests/encrypted_saved_objects_api.ts#L363).

Reply:
> How does pMap work when we have thousands of saved objects? Does it try to process everything in parallel?
Yes, it would process everything concurrently. But there is a possibility to limit it by setting `concurrency` in pMap call. I put it to 50(usual number we use async operations).

I've measured time it takes: ~ 158.960868999362ms for 100 SO with attributes that get decrypted when Kibana is relatively unloaded, and can take more time under heavy load(~ 500ms)

It takes 100 SO in batch, because I request 100 SO per page through this API

```typescript
    const rulesFinder =
      await this.encryptedSavedObjectsClient.createPointInTimeFinderAsInternalUser<RawRule>({
        filter,
        type: 'alert',
        perPage: 100,
        ...(this.namespace ? { namespaces: [this.namespace] } : undefined),
      });
```

If I understand correctly this service.decryptAttributes is non-blocking? Is that correct?

>What is the consumer (alerting) expectation if some of the saved objects cannot be decrypted (e.g. wrong encryption key)? How does API behave in this case? I think it makes sense to have a bunch of integration tests for this new API similar (but not limited to) to what we have for [getDecryptedAsInternalUser](https://github.com/elastic/kibana/blob/main/x-pack/test/encrypted_saved_objects_api_integration/tests/encrypted_saved_objects_api.ts#L363).

Expectation is similar to failure in case of using  `getDecryptedAsInternalUser`: if expected decoded property(apiKey in our case) is empty, we will just not put this apiKey for further invalidation.
Thanks for pointing out integration tests for existing method. I will look to add some.

Reply:
> Yes, it would process everything concurrently. But there is a possibility to limit it by setting concurrency in pMap call. I put it to 50(usual number we use async operations).

Got it, thanks. 50 sounds like a reasonable number to me.

> I've measured time it takes: ~ 158.960868999362ms for 100 SO with attributes that get decrypted when Kibana is relatively unloaded, and can take more time under heavy load(~ 500ms)

Thanks for measuring. I was mostly worried about how much stress it can put on Kibana with many concurrent tasks since decryption can be quite CPU intensive. Limiting concurrency "level" to 50 should alleviate this concern though.

> If I understand correctly this service.decryptAttributes is non-blocking? Is that correct?

Yes, that's correct.

> Expectation is similar to failure in case of using getDecryptedAsInternalUser: if expected decoded property(apiKey in our case) is empty, we will just not put this apiKey for further invalidation.

If I remember correctly `getDecryptedAsInternalUser` will throw if it cannot decrypt any of the attributes, so that consumer can know that decryption for the specific SO failed.

But in case of `createPointInTimeFinderAsInternalUser`, what would happen if decryption fails only for _some_ SOs? Will `createPointInTimeFinderAsInternalUser` throw or will "failed" SOs just be excluded from the resulting list of SOs? If excluded, then how would consumer know about these failed SOs?

Reply:
> I think it makes sense to have a bunch of integration tests for this new API similar (but not limited to) to what we have for [getDecryptedAsInternalUser](https://github.com/elastic/kibana/blob/main/x-pack/test/encrypted_saved_objects_api_integration/tests/encrypted_saved_objects_api.ts#L363).

implemented

> But in case of createPointInTimeFinderAsInternalUser, what would happen if decryption fails only for some SOs? Will createPointInTimeFinderAsInternalUser throw or will "failed" SOs just be excluded from the resulting list of SOs? If excluded, then how would consumer know about these failed SOs?

So, I've changed here logic a bit, thanks for pointing out that in some cases function decryptAttributes can throw.
To mitigate this and still return SO in list of saved objects and error , I wrapped it in try/catch, if it fails I enrich SO with error and also strip attributes.
```typescript
               try {
                  return {
                    ...savedObject,
                    attributes: (await service.decryptAttributes(
                      descriptor,
                      savedObject.attributes as Record<string, unknown>
                    )) as T,
                  };
                } catch (error) {
                  // catch error and enrich SO with it, return stripped attributes. Then consumer of API can decide either proceed
                  // with only unsecured properties or stop when error happens
                  const { attributes: strippedAttrs } = await service.stripOrDecryptAttributes(
                    descriptor,
                    savedObject.attributes as Record<string, unknown>
                  );
                  return {
                    ...savedObject,
                    attributes: strippedAttrs as T,
                    error: { ...error, message: `Decryption error: "${error.message}"` },
                  };
                }

```

Initially, I thought maybe I have to use `stripOrDecryptAttributes` instead of `decryptAttributes`, but I've noticed in tests: `stripOrDecryptAttributes` doesn't return all properties, hence this implementation.
For reference, loggings from functional test case `'#createPointInTimeFinderDecryptedAsInternalUser decrypts and returns all attributes'`

```bash
 decryptAttributes {
   publicProperty: 'R@)Jo#y6',
   publicPropertyStoredEncrypted: 'ahx$qhD]*RM]^f',
   privateProperty: '2D5tF)Y8Ol4#XG',
   publicPropertyExcludedFromAAD: '[x$4m82T'
 }


 stripOrDecryptAttributes {
   attributes: {
     publicProperty: 'R@)Jo#y6',
     publicPropertyExcludedFromAAD: '[x$4m82T',
     publicPropertyStoredEncrypted: 'ahx$qhD]*RM]^f'
   }
 }
```